File: core/src/main/scala/org/trustedanalytics/sparktk/frame/internal/ops/classificationmetrics/MultiClassClassificationMetrics.scala
/**
 *  Copyright (c) 2015 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package org.trustedanalytics.sparktk.frame.internal.ops.classificationmetrics

import org.trustedanalytics.sparktk.frame.internal.rdd.{ ScoreAndLabel, FrameRdd }
import scala.reflect.ClassTag
import org.apache.spark.rdd.RDD
import org.apache.commons.lang.StringUtils
import org.trustedanalytics.sparktk.frame.internal.{ FrameState, FrameSummarization, BaseFrame }

trait MultiClassClassificationMetricsSummarization extends BaseFrame {

  def multiClassClassificationMetrics(labelColumn: String,
                                      predColumn: String,
                                      beta: Double = 1.0,
                                      frequencyColumn: Option[String]): ClassificationMetricValue = {
    execute(MultiClassClassificationMetrics(labelColumn, predColumn, beta, frequencyColumn))
  }
}

/**
 * Model statistics of accuracy, precision, and others.
 *
 * @param labelColumn The name of the column containing the correct label for each instance.
 * @param predColumn The name of the column containing the predicted label for each instance.
 * @param beta This is the beta value to use for :math:`F_{ \beta}` measure (default F1 measure is
 *             computed); must be greater than zero. Default is 1.
 * @param frequencyColumn The name of an optional column containing the frequency of observations.
 */
case class MultiClassClassificationMetrics(labelColumn: String,
                                           predColumn: String,
                                           beta: Double,
                                           frequencyColumn: Option[String]) extends FrameSummarization[ClassificationMetricValue] {
  require(StringUtils.isNotEmpty(labelColumn), "label column is required")
  require(StringUtils.isNotEmpty(predColumn), "predict column is required")
  require(beta >= 0, "invalid beta value for f measure. Should be greater than or equal to 0")

  override def work(state: FrameState): ClassificationMetricValue = {
    ClassificationMetricsFunctions.multiclassClassificationMetrics(
      state,
      labelColumn,
      predColumn,
      beta,
      frequencyColumn
    )
  }
}

/**
 * Model Accuracy, Precision, Recall, FMeasure, Confusion matrix for multi-class
 */
class MultiClassMetrics[T: ClassTag](labelPredictRdd: RDD[ScoreAndLabel[T]],
                                     beta: Double = 1) extends Serializable {

  def this(frameRdd: FrameRdd,
           labelColumn: String,
           predictColumn: String,
           beta: Double = 1,
           frequencyColumn: Option[String] = None) {
    this(frameRdd.toScoreAndLabelRdd[T](labelColumn, predictColumn, frequencyColumn), beta)
  }

  labelPredictRdd.cache()

  lazy val countsByLabel = labelPredictRdd.map(scoreAndLabel => {
    (scoreAndLabel.label, scoreAndLabel.frequency)
  }).reduceByKey(_ + _).collectAsMap()

  lazy val countsByScore = labelPredictRdd.map(scoreAndLabel => {
    (scoreAndLabel.score, scoreAndLabel.frequency)
  }).reduceByKey(_ + _).collectAsMap()

  lazy val truePositivesByLabel = labelPredictRdd.map(scoreAndLabel => {
    val truePositives = if (ClassificationMetricsFunctions.compareValues(scoreAndLabel.label, scoreAndLabel.score)) {
      scoreAndLabel.frequency
    }
    else {
      0L
    }
    (scoreAndLabel.label, truePositives)
  }).reduceByKey(_ + _).collectAsMap()

  lazy val falsePositivesByScore = labelPredictRdd.map(scoreAndLabel => {
    val falsePositives = if (!ClassificationMetricsFunctions.compareValues(scoreAndLabel.label, scoreAndLabel.score)) {
      scoreAndLabel.frequency
    }
    else {
      0L
    }
    (scoreAndLabel.score, falsePositives)
  }).reduceByKey(_ + _).collectAsMap()

  lazy val totalCount = countsByLabel.map { case (label, count) => count }.sum

  lazy val totalTruePositives = truePositivesByLabel.map { case (label, count) => count }.sum

  /**
   * Compute precision for label
   *
   * Precision = true positives/(true positives + false positives)
   *
   * @param label Class label
   * @return precision
   */
  def precision(label: T): Double = {
    val truePos = truePositivesByLabel.getOrElse(label, 0L)
    val falsePos = falsePositivesByScore.getOrElse(label, 0L)

    (truePos + falsePos) match {
      case 0 => 0d
      case totalPos => truePos / totalPos.toDouble
    }
  }

  /**
   * Compute recall for label
   *
   * Recall = true positives/(true positives + false negatives)
   * @param label Class label
   * @return recall
   */
  def recall(label: T): Double = {
    val truePos = truePositivesByLabel.getOrElse(label, 0L)
    val labelCount = countsByLabel.getOrElse(label, 0L)

    labelCount match {
      case 0 => 0d
      case _ => truePos / labelCount.toDouble
    }
  }

  /**
   * Compute f-measure for label
   *
   * Fmeasure is weighted average of precision and recall
   * @param label Class label
   * @return F-measure
   */
  def fmeasure(label: T): Double = {
    val labelPrecision = precision(label)
    val labelRecall = recall(label)

    (math.pow(beta, 2) * labelPrecision) + labelRecall match {
      case 0 => 0
      case _ => (1 + math.pow(beta, 2)) * ((labelPrecision * labelRecall) / ((math.pow(beta, 2) * labelPrecision) + labelRecall))
    }
  }

  /**
   * Compute precision weighted by label occurrence
   */
  def weightedPrecision(): Double = {
    if (totalCount > 0) {
      countsByLabel.map {
        case (label, labelCount) =>
          labelCount * precision(label)
      }.sum / totalCount.toDouble
    }
    else 0d
  }

  /**
   * Compute recall weighted by label occurrence
   */
  def weightedRecall(): Double = {
    if (totalCount > 0) {
      countsByLabel.map {
        case (label, labelCount) =>
          labelCount * recall(label)
      }.sum / totalCount.toDouble
    }
    else 0d
  }

  /**
   * Compute f-measure weighted by label occurrence
   */
  def weightedFmeasure(): Double = {
    if (totalCount > 0) {
      countsByLabel.map {
        case (label, labelCount) =>
          labelCount * fmeasure(label)
      }.sum / totalCount.toDouble
    }
    else 0d
  }

  /**
   * Compute model accuracy
   */
  def accuracy(): Double = {
    if (totalCount > 0) {
      totalTruePositives / totalCount.toDouble
    }
    else 0d
  }

  /**
   * Returns string representation of the specified value
   */
  private def valueToString(value: Any): String = {
    if (value == null)
      "None"
    else
      value.toString
  }

  /**
   * Compute confusion matrix for labels
   */
  def confusionMatrix(): ConfusionMatrix = {
    val predictionSummary = labelPredictRdd.map(scoreAndLabel => {
      ((scoreAndLabel.score, scoreAndLabel.label), scoreAndLabel.frequency)
    }).reduceByKey(_ + _).collectAsMap()

    val rowLabels = predictionSummary.map { case ((score, label), frequency) => valueToString(label) }.toSet.toList.sorted
    val colLabels = predictionSummary.map { case ((score, label), frequency) => valueToString(score) }.toSet.toList.sorted
    val matrix = ConfusionMatrix(rowLabels, colLabels)
    predictionSummary.foreach {
      case ((score, label), frequency) =>
        matrix.set(valueToString(score), valueToString(label), frequency)
    }
    matrix
  }
}

File: python/sparktk/frame/ops/multiclass_classification_metrics.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

from classification_metrics_value import ClassificationMetricsValue

def multiclass_classification_metrics(self, label_column, pred_column, beta=1.0, frequency_column=None):
    """
    Statistics of accuracy, precision, and others for a multi-class classification model.

    Parameters:

    :param label_column: (str) The name of the column containing the correct label for each instance.
    :param pred_column: (str) The name of the column containing the predicted label for each instance.
    :param beta: (Optional[int]) This is the beta value to use for :math:`F_{ \beta}` measure (default F1 measure is computed);
                must be greater than zero. Defaults is 1.
    :param frequency_column: (Optional[str]) The name of an optional column containing the frequency of observations.
    :return: (ClassificationMetricsValue) The data returned is composed of multiple components:<br>
            &lt;object&gt;.accuracy : double<br>
            &lt;object&gt;.confusion_matrix : table<br>
            &lt;object&gt;.f_measure : double<br>
            &lt;object&gt;.precision : double<br>
            &lt;object&gt;.recall : double

   Calculate the accuracy, precision, confusion_matrix, recall and :math:`F_{ \beta}` measure for a
   classification model.

   *   The **f_measure** result is the :math:`F_{ \beta}` measure for a
       classification model.
       The :math:`F_{ \beta}` measure of a binary classification model is the
       harmonic mean of precision and recall.
       If we let:

       * beta :math:`\equiv \beta`,
       * :math:`T_{P}` denotes the number of true positives,
       * :math:`F_{P}` denotes the number of false positives, and
       * :math:`F_{N}` denotes the number of false negatives

       then:

       .. math::

           F_{ \beta} = (1 + \beta ^ 2) * \frac{ \frac{T_{P}}{T_{P} + F_{P}} * \
           \frac{T_{P}}{T_{P} + F_{N}}}{ \beta ^ 2 * \frac{T_{P}}{T_{P} + \
           F_{P}}  + \frac{T_{P}}{T_{P} + F_{N}}}

       The :math:`F_{ \beta}` measure for a multi-class classification model is
       computed as the weighted average of the :math:`F_{ \beta}` measure for
       each label, where the weight is the number of instances of each label.
       The determination of binary vs. multi-class is automatically inferred
       from the data.

   *   For multi-class classification models, the **recall** measure is computed as
       the weighted average of the recall for each label, where the weight is
       the number of instances of each label.
       The determination of binary vs. multi-class is automatically inferred
       from the data.

   *   For multi-class classification models, the **precision** measure is computed
       as the weighted average of the precision for each label, where the
       weight is the number of instances of each label.
       The determination of binary vs. multi-class is automatically inferred
       from the data.

   *   The **accuracy** of a classification model is the proportion of
       predictions that are correctly identified.
       If we let :math:`T_{P}` denote the number of true positives,
       :math:`T_{N}` denote the number of true negatives, and :math:`K` denote
       the total number of classified instances, then the model accuracy is
       given by: :math:`\frac{T_{P} + T_{N}}{K}`.

   *   The **confusion_matrix** result is a confusion matrix for a
       classifier model, formatted for human readability.

   Examples
   --------
   Consider Frame *my_frame*, which contains the data

       <hide>
       >>> s = [('a', str),('b', int),('labels', int),('predictions', int)]
       >>> rows = [["red", 1, 0, 0], ["blue", 3, 1, 0],["green", 1, 0, 0],["green", 0, 1, 1],["red", 0, 5, 4]]
       >>> my_frame = tc.frame.create(rows, s)
       -etc-

       </hide>
       >>> my_frame.inspect()
        [#]  a      b  labels  predictions
        ==================================
        [0]  red    1       0            0
        [1]  blue   3       1            0
        [2]  green  1       0            0
        [3]  green  0       1            1
        [4]  red    0       5            4

       >>> cm = my_frame.multiclass_classification_metrics('labels', 'predictions')
       <progress>

       >>> cm.f_measure
       0.5866666666666667

       >>> cm.recall
       0.6

       >>> cm.accuracy
       0.6

       >>> cm.precision
       0.6666666666666666

       >>> cm.confusion_matrix
                         Predicted_0  Predicted_1  Predicted_4
        Actual_0            2            0            0
        Actual_1            1            1            0
        Actual_5            0            0            1

    """
    return ClassificationMetricsValue(self._tc, self._scala.multiClassClassificationMetrics(label_column,
                                      pred_column,
                                      float(beta),
                                      self._tc.jutils.convert.to_scala_option(frequency_column)))